spring-integration-jpa/src/main/java/org/springframework/integration/jpa/core/JpaExecutor.java

```diff
@@ -628,4 +362,267 @@ public void setMaxNumberOfResults(int maxNumberOfResults) {
+				if (this.deleteInBatch) {
+					this.jpaOperations.deleteInBatch((Iterable<?>) payload);
+				}
+				else {
+					for (Object entity : (Iterable<?>) payload) {
+						this.jpaOperations.delete(entity);
+					}
+				}
+			}
+			else {
+				this.jpaOperations.delete(payload);
+			}
+
+			if (this.flush) {
+				this.jpaOperations.flush();
+			}
+		}
+		return payload;
+	}
+
+	protected List<?> doPoll(ParameterSource jpaQLParameterSource, int firstResult, int maxNumberOfResults) {
+		List<?> payload;
+		if (this.jpaQuery != null) {
+			payload =
+					this.jpaOperations.getResultListForQuery(this.jpaQuery, jpaQLParameterSource,
+							firstResult, maxNumberOfResults);
+		}
+		else if (this.nativeQuery != null) {
+			payload =
+					this.jpaOperations.getResultListForNativeQuery(this.nativeQuery, this.entityClass,
+							jpaQLParameterSource, firstResult, maxNumberOfResults);
+		}
+		else if (this.namedQuery != null) {
+			payload =
+					this.jpaOperations.getResultListForNamedQuery(this.namedQuery, jpaQLParameterSource,
+							firstResult, maxNumberOfResults);
+		}
+		else if (this.entityClass != null) {
+			payload = this.jpaOperations.getResultListForClass(this.entityClass, firstResult, maxNumberOfResults);
+		}
+		else {
+			throw new IllegalStateException("For the polling operation, one of "
+					+ "the following properties must be specified: "
+					+ "query, namedQuery or entityClass.");
+		}
+		return payload;
+	}
+
+	private int getFirstResult(final Message<?> requestMessage) {
+		return evaluateExpressionForNumericResult(requestMessage, this.firstResultExpression);
+	}
+
+	private int evaluateExpressionForNumericResult(final Message<?> requestMessage, Expression expression) {
+		int evaluatedResult = 0;
+		if (expression != null) {
+			Object evaluationResult = expression.getValue(this.evaluationContext, requestMessage);
+			if (evaluationResult != null) {
+				if (evaluationResult instanceof Number) {
+					evaluatedResult = ((Number) evaluationResult).intValue();
+				}
+				else if (evaluationResult instanceof String) {
+					try {
+						evaluatedResult = Integer.parseInt((String) evaluationResult);
+					}
+					catch (NumberFormatException e) {
+						throw new IllegalArgumentException(
+								"Value " + evaluationResult + " passed as cannot be " +
+										"parsed to a number, expected to be numeric");
+					}
+				}
+				else {
+					throw new IllegalArgumentException("Expected the value to be a Number" +
+							" got " + evaluationResult.getClass().getName());
+				}
+			}
+		}
+		return evaluatedResult;
+	}
+
+	private ParameterSource determineParameterSource(final Message<?> requestMessage) {
+		ParameterSource parameterSource;
+		if (this.usePayloadAsParameterSource) {
+			parameterSource = this.parameterSourceFactory.createParameterSource(requestMessage.getPayload());
+		}
+		else {
+			parameterSource = this.parameterSourceFactory.createParameterSource(requestMessage);
+		}
+		return parameterSource;
+	}
+
```

Comment:
Doh... Sorry, I missed the fact that it is going to be hard to review.
The other major change is `switch` for the enum instead of `if...else`.

Also I changed the logic in the `poll()` to this:
```
else if (requestMessage != null) {
	throw new MessagingException(requestMessage,
					"The Jpa operation returned more than 1 result for expectSingleResult mode.");
}
else {
	throw new MessagingException(
					"The Jpa operation returned more than 1 result for expectSingleResult mode.");
}
```

Since my IDEA complains that `requestMessage` may be `null`. I believe SonarQube has the same concern in code smell section.

Other refactoring was about moving setters before `afterPropertiesSet()`, reordering methods which use each other, removing `this.` from methods.

Reply:
We should add `@Nullable` to that parameter - even if we haven't yet enabled `NonNullApi` for this package.